File: flutter_app/lib/presentation/client/create_job_view_model.dart
import 'package:flutter/material.dart';
import 'package:firebase_auth/firebase_auth.dart' as auth;
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:uuid/uuid.dart';

import '../../domain/model/data_models.dart';

class CreateJobUiState {
  final bool isLoading;
  final bool isSuccess;
  final String? errorMessage;
  final List<ElderProfile> elders;
  final double selectedCaregiverRate;

  CreateJobUiState({
    this.isLoading = false,
    this.isSuccess = false,
    this.errorMessage,
    this.elders = const [],
    this.selectedCaregiverRate = 0.0,
  });

  CreateJobUiState copyWith({
    bool? isLoading,
    bool? isSuccess,
    String? errorMessage,
    List<ElderProfile>? elders,
    double? selectedCaregiverRate,
  }) {
    return CreateJobUiState(
      isLoading: isLoading ?? this.isLoading,
      isSuccess: isSuccess ?? this.isSuccess,
      errorMessage: errorMessage ?? this.errorMessage,
      elders: elders ?? this.elders,
      selectedCaregiverRate: selectedCaregiverRate ?? this.selectedCaregiverRate,
    );
  }
}

class CreateJobViewModel extends ChangeNotifier {
  final auth.FirebaseAuth _auth = auth.FirebaseAuth.instance;
  final FirebaseFirestore _firestore = FirebaseFirestore.instance;

  CreateJobUiState _uiState = CreateJobUiState();
  CreateJobUiState get uiState => _uiState;

  CreateJobViewModel() {
    fetchElders();
  }

  void _setUiState(CreateJobUiState state) {
    _uiState = state;
    notifyListeners();
  }

  void fetchElders() {
    final userId = _auth.currentUser?.uid;
    if (userId == null) return;
    
    _firestore.collection("elders")
        .where("clientId", isEqualTo: userId)
        .snapshots()
        .listen((snapshot) {
      final eldersList = snapshot.docs.map((doc) => ElderProfile.fromMap(doc.data())).toList();
      _setUiState(_uiState.copyWith(elders: eldersList));
    });
  }

  Future<void> fetchCaregiverRate(String? caregiverId) async {
    if (caregiverId == null || caregiverId.isEmpty) return;
    try {
      final doc = await _firestore.collection("users").doc(caregiverId).get();
      if (doc.exists) {
        final user = User.fromMap(doc.data()!);
        _setUiState(_uiState.copyWith(selectedCaregiverRate: user.hourlyRate));
      }
    } catch (e) {
      // Ignore error, fallback to 0.0
    }
  }

  Future<void> createJobRequest({
    required double lat,
    required double lng,
    required String description,
    required bool isEmergency,
    required String elderId,
    String? selectedCaregiverId,
    required int scheduledTimeMs,
    required String careCategory,
    required String serviceType,
    required String durationType,
    required List<String> requestedTasks,
    required int estimatedHours,
  }) async {
    _setUiState(_uiState.copyWith(isLoading: true, errorMessage: null, isSuccess: false));
    try {
      final clientId = _auth.currentUser?.uid;
      if (clientId == null) return;
      
      final jobId = const Uuid().v4();
      final totalAmount = estimatedHours * _uiState.selectedCaregiverRate;

      final newBooking = Booking(
        id: jobId,
        clientId: clientId,
        caregiverId: selectedCaregiverId ?? '',
        elderId: elderId,
        status: (selectedCaregiverId != null && selectedCaregiverId.isNotEmpty) ? "PENDING" : "BROADCASTED",
        locationLat: lat,
        locationLng: lng,
        jobDescription: description,
        isEmergency: isEmergency,
        requestedTime: DateTime.now().millisecondsSinceEpoch,
        scheduledTime: scheduledTimeMs,
        careCategory: careCategory,
        serviceType: serviceType,
        durationType: durationType,
        requestedTasks: requestedTasks,
        estimatedHours: estimatedHours,
        hourlyRate: _uiState.selectedCaregiverRate,
        totalAmount: totalAmount,
        isSubscriptionBooking: false,
        planName: '',
        allowedCaregivers: const [],
        isPaid: false,
      );

      await _firestore.collection("bookings").doc(jobId).set(newBooking.toMap());
      _setUiState(_uiState.copyWith(isLoading: false, isSuccess: true));
    } catch (e) {
      _setUiState(_uiState.copyWith(isLoading: false, errorMessage: e.toString()));
    }
  }

  void resetSuccessState() {
    _setUiState(_uiState.copyWith(isSuccess: false));
  }
}

File: app/src/main/java/com/kina/care/presentation/client/CreateJobViewModel.kt
package com.kina.care.presentation.client

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.firestore.FirebaseFirestore
import com.kina.care.domain.model.Booking
import com.kina.care.domain.model.ElderProfile
import com.kina.care.domain.model.User
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.tasks.await
import java.util.UUID

data class CreateJobUiState(
    val isLoading: Boolean = false,
    val isSuccess: Boolean = false,
    val errorMessage: String? = null,
    val elders: List<ElderProfile> = emptyList(),
    val selectedCaregiverRate: Double = 0.0 // 🌟 අලුත්: Caregiver ගේ ගාස්තුව පෙන්වීමට
)

class CreateJobViewModel : ViewModel() {
    private val firestore = FirebaseFirestore.getInstance()
    private val auth = FirebaseAuth.getInstance()

    private val _uiState = MutableStateFlow(CreateJobUiState())
    val uiState: StateFlow<CreateJobUiState> = _uiState.asStateFlow()

    init { fetchElders() }

    private fun fetchElders() {
        val userId = auth.currentUser?.uid ?: return
        firestore.collection("elders")
            .whereEqualTo("clientId", userId)
            .addSnapshotListener { snapshot, error ->
                if (error != null) return@addSnapshotListener
                if (snapshot != null) {
                    val eldersList = snapshot.documents.mapNotNull { it.toObject(ElderProfile::class.java) }
                    _uiState.value = _uiState.value.copy(elders = eldersList)
                }
            }
    }

    // 🌟 අලුත්: තෝරාගත් Caregiver ගේ පැයක ගාස්තුව ලබා ගැනීම
    fun fetchCaregiverRate(caregiverId: String?) {
        if (caregiverId == null) return
        viewModelScope.launch {
            try {
                val doc = firestore.collection("users").document(caregiverId).get().await()
                if (doc.exists()) {
                    val user = doc.toObject(User::class.java)
                    val rate = user?.hourlyRate ?: 0.0
                    _uiState.value = _uiState.value.copy(selectedCaregiverRate = rate)
                }
            } catch (e: Exception) {
                // Ignore error, fallback to 0.0
            }
        }
    }

    fun createJobRequest(
        lat: Double, lng: Double, description: String, isEmergency: Boolean,
        elderId: String, selectedCaregiverId: String? = null, scheduledTimeMs: Long = 0L,
        careCategory: String, serviceType: String, durationType: String,
        requestedTasks: List<String>, estimatedHours: Int // 🌟 අලුත්: පැය ගණන ලබා ගැනීම
    ) {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoading = true)
            try {
                val clientId = auth.currentUser?.uid ?: return@launch
                val jobId = UUID.randomUUID().toString()

                // මුදල ගණනය කිරීම (පැය ගණන * පැයක ගාස්තුව)
                val totalAmount = estimatedHours * _uiState.value.selectedCaregiverRate

                val newBooking = Booking(
                    id = jobId,
                    clientId = clientId,
                    caregiverId = selectedCaregiverId,
                    elderId = elderId,
                    status = if (selectedCaregiverId != null) "PENDING" else "BROADCASTED",
                    locationLat = lat,
                    locationLng = lng,
                    jobDescription = description,
                    isEmergency = isEmergency,
                    requestedTime = System.currentTimeMillis(),
                    scheduledTime = scheduledTimeMs,
                    careCategory = careCategory,
                    serviceType = serviceType,
                    durationType = durationType,
                    requestedTasks = requestedTasks,
                    estimatedHours = estimatedHours, // 🌟 Save the hours
                    hourlyRate = _uiState.value.selectedCaregiverRate,
                    totalAmount = totalAmount // 🌟 Save the calculated total amount
                )

                firestore.collection("bookings").document(jobId).set(newBooking).await()
                _uiState.value = _uiState.value.copy(isLoading = false, isSuccess = true)
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(isLoading = false, errorMessage = e.localizedMessage)
            }
        }
    }
}